src/gtest/test_pow.cpp

```diff
@@ -0,0 +1,62 @@
+#include <gtest/gtest.h>
+
+#include "chain.h"
+#include "chainparams.h"
+#include "pow.h"
+#include "random.h"
+
+TEST(PoW, DifficultyAveraging) {
+    SelectParams(CBaseChainParams::MAIN);
+    const Consensus::Params& params = Params().GetConsensus();
+    size_t lastBlk = 2*params.nPowAveragingWindow;
+    size_t firstBlk = lastBlk - params.nPowAveragingWindow;
+
+    // Start with blocks evenly-spaced and equal difficulty
+    std::vector<CBlockIndex> blocks(lastBlk+1);
+    for (int i = 0; i <= lastBlk; i++) {
+        blocks[i].pprev = i ? &blocks[i - 1] : nullptr;
+        blocks[i].nHeight = i;
+        blocks[i].nTime = 1269211443 + i * params.nPowTargetSpacing;
+        blocks[i].nBits = 0x1e7fffff; /* target 0x007fffff000... */
+        blocks[i].nChainWork = i ? blocks[i - 1].nChainWork + GetBlockProof(blocks[i - 1]) : arith_uint256(0);
+    }
+
+    // Result should be the same as if last difficulty was used
+    EXPECT_EQ(CalculateNextWorkRequired(&blocks[lastBlk],
+                                        blocks[firstBlk].GetMedianTimePast(),
+                                        params),
+              GetNextWorkRequired(&blocks[lastBlk], nullptr, params));
+    // Result should be unchanged
+    // TODO: This should be 0x1e7fffff, and just before GetNextWorkRequired()
+    //       returns, it is. Somehow it ends up off by one....
```

Comment:
I already did try, and my findings are what I documented in the comment. Just before `GetNextWorkRequired` returns the result, the result is correct. After it returned the result, the macro sees it is off-by-one. The only think I didn't check is what the result from `GetNextWorkRequired` looks like before entering the macro; logically they should be identical.